src/inmanta/ast/entity.py

```diff
@@ -508,9 +525,199 @@ def get_double_defined_exception(self, other: "Namespaced") -> "DuplicateExcepti
     def get_location(self) -> Location:
         return self.location
 
+    def pair_dataclass(self) -> None:
+        """
+        Attach the associated dataclass in the python domain
+
+        should only be called on children of std::Dataclass
+        should be called after normalization
+        """
+        assert self.normalized
+
+        # Find the dataclass name
+        namespace = self.namespace.get_full_name()
+        module_name = "inmanta_plugins." + namespace.replace("::", ".")
+        dataclass_name = module_name + "." + self.name
+
+        # Find the dataclass
+        dataclass_module = importlib.import_module(module_name)
+        dataclass_raw = getattr(dataclass_module, self.name, None)
+        if dataclass_raw is None:
+            raise DataClassMismatchException(
+                self,
+                None,
+                dataclass_name,
+                f"The dataclass {self.get_full_name()} defined at {self.location} has no corresponding python dataclass. "
+                "Dataclasses must have a python counterpart that is a frozen dataclass.",
+            )
 
-# Kept for backwards compatibility. May be dropped from iso7 onwards.
-EntityLike = Entity
+        if not dataclasses.is_dataclass(dataclass_raw):
+            raise DataClassMismatchException(
+                self,
+                None,
+                dataclass_name,
+                f"The python object {module_name}.{dataclass_raw.__name__} associated to  {self.get_full_name()} "
+                f"defined at {self.location} is not a dataclass. "
+                "Dataclasses must have a python counterpart that is a frozen dataclass.",
+            )
+        dataclass: type[DataclassProtocol] = dataclass_raw
+        if not dataclass.__dataclass_params__.frozen:
+            raise DataClassMismatchException(
+                self,
+                None,
+                dataclass_name,
+                f"The python object {module_name}.{dataclass.__name__} associated to  {self.get_full_name()} "
+                f"defined at {self.location} is not frozen. "
+                "Dataclasses must have a python counterpart that is a frozen dataclass.",
+            )
+
+        # Validate fields, collect errors
+        dc_fields = {f.name: f for f in dataclasses.fields(dataclass)}
+        dc_types = typing.get_type_hints(dataclass)
+        failures = []
+
+        for rel_or_attr_name in self.get_all_attribute_names():
+            rel_or_attr = self.get_attribute(rel_or_attr_name)
+            match rel_or_attr:
+                case inmanta.ast.attribute.RelationAttribute() as rel:
+                    # No relations except for requires and provides
+                    if not rel.entity.get_full_name() == "std::Entity":
```

Comment:
Isn't this too accepting? e.g. `MyCollector.all [0:] -- std::Entity` would be accepted here, wouldn't it?

Reply:
Ah, I (wrongfully) assumed that `relation.entity` was the other side of the relation, but it's the one the relation is defined on. So I agree, it could only happen in the scenario you describe. But we should probably still cover it, even if I don't expect anyone would ever really do something like that.

Reply:
I updated it, but it will break of you define ANY relation on std::entity, so a bit of a tough one

Reply:
> it will break of you define ANY relation on std::entity

Isn't that what we want? At least for now.

Reply:
I think so